### PR TITLE
Forward new Prometheus warnings from querier to query frontend

### PR DESCRIPTION
**Pull request by zenador**
#### What this PR does

Prometheus now shows warnings for some queries which can be received at the querier, but does not show up in the query frontend. This fixes that.

To test, you can run this query `quantile(10, up)`. A warning should already appear when you send it to the querier, and with this PR a warning should also appear when you send it to the query frontend.

#### Which issue(s) this PR fixes or relates to

N/A

#### Checklist

- [ ] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: CHANGELOG.md
Comment (charleskorn) [inline CHANGELOG.md:50]: [nit] Would this be clearer perhaps?
```suggestion
* [ENHANCEMENT] Query-frontend: return warnings generated during query evaluation. #6391
```

File: pkg/frontend/querymiddleware/codec.go
Comment (charleskorn) [inline pkg/frontend/querymiddleware/codec.go:183]: Could we do this instead? This would avoid the need to set `promWarnings` to `nil` later if it's empty.

```suggestion
	var promWarnings []string
```

File: pkg/frontend/querymiddleware/querysharding.go
Comment (charleskorn) [inline pkg/frontend/querymiddleware/querysharding.go:170]: Does this mean the positions reported in the warning won't match the query the user provided? If so, can we do anything to either adjust the reported positions so they do match, or remove them altogether so it's not confusing?

**Comment by charleskorn** [comment]
Also, it'd be good to add an integration test showing that warnings flow through all the components correctly, with and without features like query sharding enabled.

File: pkg/frontend/querymiddleware/querysharding.go
Comment (zenador) [reply pkg/frontend/querymiddleware/querysharding.go:170]: Yes, it won't. I'm not sure how to fix this actually. So the flow is, the user submits query A, and it gets rewritten behind the scenes into query B. So using either query A (not what is actually run) or B (not shown to the user) as the reference would be confusing. We can also pass in an empty string so the positions would be shown as invalid.

Comment (charleskorn) [reply pkg/frontend/querymiddleware/querysharding.go:170]: Could we implement our own version of `AsStrings()` that doesn't emit the positions at all? That seems like the least confusing option. 

This has also triggered two other questions for me:
* What happens if the same warning is generated by multiple shards? Would the user get the same warning multiple times? Or would we deduplicate them?
* Are there any cases where sharding the query would generate a warning? eg. if we executed the query the user wrote without sharding, there would be no warnings, but something about how it is rewritten causes a warning to be generated

Comment (zenador) [reply pkg/frontend/querymiddleware/querysharding.go:170]: Yes, that's a good idea, I'll make it so that our fork will hide the "invalid position" messages when an empty string is passed in (so that we can continue showing the positions when the query is unsharded).

It should be deduplicated as the underlying data structure for the warnings is a map.

Yes. There's one particular warning (a bit of an edge case) that should appear only for sharded queries, where the sharding makes the end results a bit imprecise, and passing them into histogram_quantile would give invalid results.

Comment (zenador) [reply pkg/frontend/querymiddleware/querysharding.go]: Updated in our fork: https://github.com/grafana/mimir-prometheus/pull/550

**Comment by zenador** [comment]
Added a unit test instead to compare the raw results vs sharded and split queries, is that okay?

File: pkg/frontend/querymiddleware/codec.go
Comment (charleskorn) [inline pkg/frontend/querymiddleware/codec.go:183]: Using a `map[string]struct{}` here should be faster and use less memory: https://itnext.io/set-in-go-map-bool-and-map-struct-performance-comparison-5315b4b107b has some benchmarks and https://stackoverflow.com/a/22770744 has some other data that explains why.

File: pkg/frontend/querymiddleware/querysharding_test.go
Comment (charleskorn) [inline pkg/frontend/querymiddleware/querysharding_test.go:1647]: Is there a reason to use `make(..., 0)` here rather than declaring `storageSeries` following the same pattern as for `floats` below (ie. `var storageSeries []*promql.StorageSeries`)?

Comment (charleskorn) [reply pkg/frontend/querymiddleware/querysharding_test.go:1647]: Alternatively, we could use `numSeries` instead of `0` here.

**Review by charleskorn — COMMENTED** [review]
>
Added a unit test instead to compare the raw results vs sharded and split queries, is that okay?

Yep, test looks good to me.

File: pkg/frontend/querymiddleware/codec.go
Comment (zenador) [reply pkg/frontend/querymiddleware/codec.go]: Thanks for the helpful links!

File: pkg/frontend/querymiddleware/querysharding_test.go
Comment (zenador) [reply pkg/frontend/querymiddleware/querysharding_test.go]: Made them consistent and declared capacities up front.

**Comment by zenador** [comment]
We're upstreaming the [change to hide position info](https://github.com/prometheus/prometheus/pull/13031) so that'll be taken care of separate from this PR when we update mimir-prometheus vendor.

File: pkg/frontend/querymiddleware/codec.go
Comment (charleskorn) [inline pkg/frontend/querymiddleware/codec.go:184]: [nit] 
```suggestion
	var present struct{}
```

File: pkg/frontend/querymiddleware/querysharding_test.go
Comment (charleskorn) [inline pkg/frontend/querymiddleware/querysharding_test.go:1721]: [nit] We should be able to extract the `user.InjectOrgID(context.Background(), "test")` call here and below to a single context used by both calls

**Review by charleskorn — APPROVED** [review]
LGTM modulo nits and Prometheus upgrade to bring in https://github.com/prometheus/prometheus/pull/13031

File: pkg/frontend/querymiddleware/codec.go
Comment (charleskorn) [reply pkg/frontend/querymiddleware/codec.go:184]: (`empty` makes it sound like it's not there)

**Comment by zenador** [comment]
Thank you for the detailed review and helpful tips! Do we need to wait for the Prometheus update before merging?

**Comment by charleskorn** [comment]
> Thank you for the detailed review and helpful tips! Do we need to wait for the Prometheus update before merging?

I think that would be best, otherwise we'll potentially end up showing the `unknown position` messages to users if we only merge this change.

**Comment by zenador** [comment]
Pending https://github.com/grafana/mimir/pull/6489

**Pull request merged**
